File: app/src/main/java/com/mando/myweather/location/FusedLocationDataStore.java
/*
 * Copyright (c) 2018. OpenSignal.com
 */

package com.mando.myweather.location;

import android.annotation.SuppressLint;
import android.content.Context;
import android.location.Location;
import android.os.Looper;
import android.util.Log;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;

import com.google.android.gms.location.FusedLocationProviderClient;
import com.google.android.gms.location.LocationCallback;
import com.google.android.gms.location.LocationResult;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;
import com.google.android.gms.tasks.Tasks;
import com.mando.myweather.utils.AndroidPermissionChecker;
import com.mando.myweather.utils.PermissionChecker;

import java.util.concurrent.CopyOnWriteArrayList;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

public class FusedLocationDataStore implements LocationDataStore {
    public static final String TAG = "FusedLocationDataStore";

    private FusedLocationProviderClient mFusedLocationClient;

    private LocationCallback mLocationCallback;
    private DeviceLocation mLocation;
    private PermissionChecker permissionChecker;

    private CopyOnWriteArrayList<LocationDataStoreListener> mListeners = new CopyOnWriteArrayList<>();

    private FusedLocationDataStore() {
    }

    @VisibleForTesting
    void initialize(Context context) {
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(context);
        permissionChecker = new AndroidPermissionChecker(context);

        initLocationCallback();
        updateLastLocation();
    }

    private void initLocationCallback() {
        mLocationCallback = new LocationCallback() {
            @Override
            public void onLocationResult(LocationResult locationResult) {
                Log.i(TAG, "onLocationResult() called with: locationResult = [" + locationResult + "]" + " From thread: " + Thread.currentThread().getId() + " isMainThread [" + (Looper.myLooper() == Looper.getMainLooper()) + "]");

                if (locationResult == null) {
                    Log.w(TAG, "onLocationResult() received invalid locationResult: [" + locationResult + "]");
                    return;
                }

                mLocation = new DeviceLocation(locationResult.getLastLocation());
                Log.i(TAG, "onLocationResult() called with: location = [" + mLocation + "]");
                dispatchLocationChanged();
            }
        };
    }


    public DeviceLocation getLocation() {
        updateLastLocation();
        Log.d(TAG, "getLocation() returned: " + mLocation);
        return mLocation;
    }

    @SuppressLint("MissingPermission")
    @VisibleForTesting
    void updateLastLocation() {
        Log.d(TAG, "updateLastLocation() called" +
                " From thread: " + Thread.currentThread().getId() +
                " isMainThread [" + (Looper.myLooper() == Looper.getMainLooper()) + "]");

        if (!permissionChecker.getHasAnyLocationPermissions()) {
            Log.d(TAG, "No Location permission granted");
            return;
        }
        final Task<Location> lastLocationTask = mFusedLocationClient.getLastLocation();
        lastLocationTask.addOnCompleteListener(new OnCompleteListener<Location>() {
            @Override
            public void onComplete(@NonNull Task<Location> task) {
                if (lastLocationTask.isSuccessful()) {
                    final Location location = lastLocationTask.getResult();
                    if (location != null) {
                        mLocation = new DeviceLocation(location);
                    }
                }else {
                    Log.d(TAG, "Location is null");
                }
            }
        });

    }

    @VisibleForTesting
    void dispatchLocationChanged() {
        for (LocationDataStoreListener listener : mListeners) {
            listener.onLocationChanged(mLocation);
        }
    }

    public static FusedLocationDataStore getInstance(Context context) {
        final FusedLocationDataStore fusedLocationDataStore = InstanceHolder.sInstance;

        if (fusedLocationDataStore.mFusedLocationClient == null) {
            fusedLocationDataStore.initialize(context);
        }

        return fusedLocationDataStore;
    }

    private static final class InstanceHolder {
        private static final FusedLocationDataStore sInstance = new FusedLocationDataStore();
    }
}

File: app/src/main/java/com/mando/myweather/location/FusedLocationDataStore.kt
package com.mando.myweather.location

import android.annotation.SuppressLint
import android.content.Context
import android.os.Looper
import android.util.Log
import androidx.annotation.VisibleForTesting
import com.google.android.gms.location.*
import com.mando.myweather.utils.AndroidPermissionChecker
import com.mando.myweather.utils.PermissionExaminer
import timber.log.Timber

private const val DEFAULT_INTERVAL = 0L
const val TAG = "FusedLocationDataStore"

object FusedLocationDataStore : LocationDataStore {

    private var mFusedLocationClient: FusedLocationProviderClient? = null
    private var mLocation: DeviceLocation? = null
    private var mPermissionExaminer: PermissionExaminer? = null

    private val locationCallback = object : LocationCallback() {

        override fun onLocationResult(locationResult: LocationResult?) {
            Timber.w( "onLocationResult($locationResult)")

            if (locationResult == null || locationResult.lastLocation == null) {
                Timber.w("Location is null. Returning")
                return
            }
            mLocation = DeviceLocation(locationResult.lastLocation)
        }
    }

    private fun initialize(context: Context?) {
        Timber.d("initialize($context)")
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(context!!)
        mPermissionExaminer = AndroidPermissionChecker(context)
        updateLastLocation()
    }

    override val location: DeviceLocation?
        get() {
            updateLastLocation()
            Timber.d( "getLocation() returned: $mLocation")
            return mLocation
        }

    override fun requestNewLocation() {
        if (mPermissionExaminer?.hasAnyLocationPermissions == false) {
            Timber.w("Cannot request a new location as we don't have permission.")
            return
        }

        val request = createLocationRequest()
        mFusedLocationClient?.requestLocationUpdates(
            request,
            locationCallback,
            Looper.getMainLooper()
        )
    }
    private fun createLocationRequest() = when (mPermissionExaminer?.hasFineLocationPermission) {
        true -> createSingleLocationRequestHighAccuracy()
        else -> createSingleLocationRequestBalancedPower()
    }
    private fun createSingleLocationRequestHighAccuracy() =
        createLocationRequest(LocationRequest.PRIORITY_HIGH_ACCURACY)

    private fun createSingleLocationRequestBalancedPower() =
        createLocationRequest(LocationRequest.PRIORITY_BALANCED_POWER_ACCURACY)

    private fun createLocationRequest(requestPriority: Int): LocationRequest {
        val locationUpdateInterval =  DEFAULT_INTERVAL
        val locationUpdateFastestInterval =  DEFAULT_INTERVAL

        return LocationRequest().apply {
            interval = locationUpdateInterval
            fastestInterval = locationUpdateFastestInterval
            priority = requestPriority
        }
    }

    @SuppressLint("MissingPermission")
    @VisibleForTesting
    fun updateLastLocation() {
        Timber.d("updateLastLocation() called" +
                    " From thread: " + Thread.currentThread().id +
                    " isMainThread [" + (Looper.myLooper() == Looper.getMainLooper()) + "]"
        )
        if (mPermissionExaminer?.hasAnyLocationPermissions == false) {
            Timber.d("No Location permission granted")
            return
        }
        val lastLocationTask =
            mFusedLocationClient!!.lastLocation
        lastLocationTask.addOnCompleteListener {
            if (lastLocationTask.isSuccessful) {
                val location = lastLocationTask.result
                if (location != null) {
                    mLocation = DeviceLocation(location)
                }
            } else {
                Timber.d( "Location is null")
            }
        }
    }

    fun getInstance(context: Context?): FusedLocationDataStore? {
        val fusedLocationDataStore = InstanceHolder.sInstance
        if (fusedLocationDataStore.mFusedLocationClient == null) {
            fusedLocationDataStore.initialize(context)
        }
        return fusedLocationDataStore
    }

    private object InstanceHolder {
        val sInstance: FusedLocationDataStore = FusedLocationDataStore
    }
}